### PR TITLE
Pickup Checklist Page

### PR DESCRIPTION
**Pull request by seanpdoyle**
https://trello.com/c/HG23JmsS

When donations are ready to be picked up, provide a page for cyclists to
view that lists all the confirmed donations, along with their donor's
names and addresses.

### CODE REVIEW

File: app/controllers/pickup_checklists_controller.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class PickupChecklistsController < ApplicationController
+  def show
+    @pickup_checklist = PickupChecklist.new(find_scheduled_pickup)
+  end
+
+  def find_scheduled_pickup
```
Comment (geoffharcourt) [inline app/controllers/pickup_checklists_controller.rb]: Do these methods need to be `public`?

File: app/views/pickup_checklists/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<div class="content-wrapper-large content-container">
+  <h3>
+    <%= t(".header") %>
+    <%= link_to(
+      zone_scheduled_pickup_path(@pickup_checklist.zone, @pickup_checklist),
+      class: "pickup-checklist-date",
+    ) do %>
+      <%= l(@pickup_checklist.date) %>
+    <% end %>
```
Comment (geoffharcourt) [inline app/views/pickup_checklists/show.html.erb]: This is a little awkward in ERB. Could a helper method improve the display here?

(Also, is there another way to do the path without having to pass two items in? Is there a `pickups` path that can figure out the path just from the checklist without the zone?)

File: spec/factories.rb [hunk 3]
```diff
@@ -12,10 +12,20 @@
       confirmed true
     end
 
+    trait :confirmed_then_declined do
+      confirmed_at 1.week.ago
+      declined_at 1.day.ago
+    end
```
Comment (geoffharcourt) [inline spec/factories.rb]: This is an awesome trait design.

File: spec/factories.rb [hunk 4]
```diff
@@ -91,6 +102,10 @@
       with_location
     end
 
+    factory :cyclist do
+      admin
+    end
+
     trait :admin do
       admin true
     end
```
Comment (geoffharcourt) [inline spec/factories.rb]: I think you can make `:admin` a subfactory with `:cyclist` then being an alias here if you like the way that looks better. I often find myself using aliases so the declarations in specs are shorter.

File: app/controllers/pickup_checklists_controller.rb (hunk 1 above)
Comment (seanpdoyle) [inline app/controllers/pickup_checklists_controller.rb]: nope.

File: spec/models/donation_spec.rb [hunk 5]
```diff
@@ -21,6 +22,24 @@
...
+      reconfirmed_donation = create(:donation, declined_at: 1.week.ago, confirmed_at: 1.day.ago)
+      create(:donation, confirmed_at: 1.week.ago, declined_at: 1.day.ago)
+      create(:donation, :declined)
+      create(:donation, :pending)
+
+      confirmed = Donation.confirmed
+      donors = confirmed.map(&:donor)
+
+      expect(donors.map(&:name)).to eq([
+        confirmed_donation.donor.name,
```
Comment (geoffharcourt) [inline spec/models/donation_spec.rb]: This current Hound default drives me batty. You can fix it by putting the array bracket one line down.

File: app/models/donation.rb [hunk 6]
```diff
@@ -17,6 +18,15 @@ class Donation < ActiveRecord::Base
...
+    where(<<-SQL)
+      confirmed_at IS NOT NULL AND (
+        declined_at IS NULL OR
+        confirmed_at > declined_at
+      )
+    SQL
+  end
+
   def self.current
     joins(:scheduled_pickup).merge(ScheduledPickup.current)
```
Comment (geoffharcourt) [inline app/models/donation.rb]: I need to use AR `.merge` more often.

**Comment by geoffharcourt** [comment]
There's a couple possibilities for complex link-tag and factory reorganization, but you're in good shape here overall.

File: app/views/pickup_checklists/show.html.erb (hunk 2 above)
Comment (seanpdoyle) [inline app/views/pickup_checklists/show.html.erb]: @geoffharcourt are you suggesting a `/pickups/:id` path whose controller handles the redirection?

**Pull request merged**